infrastructure/keymanager/src/mnemonic.rs

```diff
@@ -0,0 +1,502 @@
+    let bytes = to_bytes(mnemonic_seq)?;
+    //(MnemonicError::from(SecretKey::from_bytes(&bytes)))
+    match SecretKey::from_bytes(&bytes) {
+        Ok(k) => Ok(k),
+        Err(e) => Err(MnemonicError::from(e)),
+    }
+}
+
+/// Generates a SecretKey that represent the provided mnemonic sequence of words using the specified language
+pub fn to_secretkey_with_language(
+    mnemonic_seq: &Vec<String>,
+    language: &MnemonicLanguage,
+) -> Result<SecretKey, MnemonicError>
+{
+    let bytes = to_bytes_with_language(mnemonic_seq, language)?;
+    match SecretKey::from_bytes(&bytes) {
+        Ok(k) => Ok(k),
+        Err(e) => Err(MnemonicError::from(e)),
+    }
+}
+
+pub trait Mnemonic {
+    fn from_mnemonic(mnemonic_seq: &Vec<String>) -> Result<SecretKey, MnemonicError>;
+    fn from_mnemonic_with_language(
+        mnemonic_seq: &Vec<String>,
+        language: &MnemonicLanguage,
+    ) -> Result<SecretKey, MnemonicError>;
+    fn to_mnemonic(&self, language: &MnemonicLanguage) -> Result<Vec<String>, MnemonicError>;
+}
+
+impl Mnemonic for SecretKey {
+    /// Generates a SecretKey that represent the provided mnemonic sequence of words, the language of the mnemonic
+    /// sequence is autodetected
+    fn from_mnemonic(mnemonic_seq: &Vec<String>) -> Result<SecretKey, MnemonicError> {
+        (to_secretkey(mnemonic_seq))
+    }
+
+    /// Generates a SecretKey that represent the provided mnemonic sequence of words using the specified language
+    fn from_mnemonic_with_language(
+        mnemonic_seq: &Vec<String>,
+        language: &MnemonicLanguage,
+    ) -> Result<SecretKey, MnemonicError>
+    {
+        (to_secretkey_with_language(mnemonic_seq, language))
+    }
+
+    /// Generates a mnemonic sequence of words from the provided secret key
+    fn to_mnemonic(&self, language: &MnemonicLanguage) -> Result<Vec<String>, MnemonicError> {
+        (from_secretkey(self, language))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crypto::{common::ByteArray, keys::SecretKeyFactory, ristretto::RistrettoSecretKey as SecretKey};
+    use mnemonic;
+    use rand;
+
+    #[test]
+    fn test_check_wordlists_sorted() {
+        for i in 0..2047 {
+            if (MNEMONIC_CHINESE_SIMPLIFIED_WORDS[i] > MNEMONIC_CHINESE_SIMPLIFIED_WORDS[i + 1]) ||
+                (MNEMONIC_ENGLISH_WORDS[i] > MNEMONIC_ENGLISH_WORDS[i + 1]) ||
+                (MNEMONIC_FRENCH_WORDS[i] > MNEMONIC_FRENCH_WORDS[i + 1]) ||
+                (MNEMONIC_ITALIAN_WORDS[i] > MNEMONIC_ITALIAN_WORDS[i + 1]) ||
+                (MNEMONIC_JAPANESE_WORDS[i] > MNEMONIC_JAPANESE_WORDS[i + 1]) ||
+                (MNEMONIC_KOREAN_WORDS[i] > MNEMONIC_KOREAN_WORDS[i + 1]) ||
+                (MNEMONIC_SPANISH_WORDS[i] > MNEMONIC_SPANISH_WORDS[i + 1])
+            {
+                assert!(false);
+            }
+        }
+        assert!(true);
+    }
+
+    #[test]
+    fn test_language_detection() {
+        // Test valid Mnemonic words
```

Comment:
Are there any words that are in more than one list? e.g. in English and Spanish, that could break autodetection?

Reply:
The different wordlists don't have shared words except the Chinese Traditional and Chinese Simplified sets. I removed Chinese Traditional until a better mechanism can be found to distinguish between these two languages.